src/arch/xtensa/hal/atomics.c

```diff
@@ -0,0 +1,15 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2021 Google LLC. All rights reserved.
+//
+
+int _xt_atomic_compare_exchange_4(unsigned int *_ptr, unsigned int _exp, unsigned int _val);
+
+int _xt_atomic_compare_exchange_4(unsigned int *_ptr, unsigned int _exp, unsigned int _val)
+{
+    int old_ptr = *_ptr;
+    if (old_ptr == _exp) {
+	*_ptr = _val;
+    }
+    return old_ptr;
```

Comment:
xtensa has an atomic ISA that is used already exposed via some atomic APIs, could we reuse the existing xtensa API code and wrap to the calling API expected by C++ ?
@andyross fyi - do we have atomic compare/exchange API for xtensa in Zephyr ? looks like we need it.

Reply:
We do, there's an atomic.h header that exposes a Zephyr-specific API.  Though on most architectures, including Xtensa, it just falls through to the existing gcc intrinsics.  Unfortunately xcc doesn't support that, so for Cadence toolchain builds we have our own inlines, they look like this: https://github.com/zephyrproject-rtos/zephyr/blob/main/include/arch/xtensa/atomic_xtensa.h

All that said... the code being discussed right here isn't atomic!  This is just C code, and if something comes in between lines 10 and 12 that modifies *_ptr, this will fail to see it.  You can't do this with simple logic[1], you need to use the S32C1I instruction and some care.

[1] Actually that's not quite right, there are techniques for mutual exclusion that don't rely on hardware compare-and-swap, but they're all very high overhead and mostly just something you see discussed in computer science textbooks.